File: lib/src/features/tree/domain/lineage.dart
import '../../relate/kinship.dart';
import 'family_graph.dart';

/// The highlight sets for "View lineage" of one person.
///
/// [immediate] highlights parents, siblings, spouse(s) and children. With
/// [full] it extends to all ancestors and all descendants. The edge sets are
/// expressed as member ids so the painter can decide which connectors to
/// highlight without knowing the graph.
class Lineage {
  const Lineage({
    required this.selectedId,
    required this.members,
    required this.labels,
    required this.unionMembers,
    required this.descentChildIds,
    required this.descentParentIds,
  });

  final String selectedId;

  /// Selected person + every highlighted relative (for ring/badge).
  final Set<String> members;

  /// relativeId -> capitalized relationship label ("Father", "Son", …).
  final Map<String, String> labels;

  /// A spouse bar highlights if either end is in this set.
  final Set<String> unionMembers;

  /// A child's up-link (to its parents) highlights if the child is in this set.
  final Set<String> descentChildIds;

  /// A parent's down-links (to its children) highlight if the parent is here.
  final Set<String> descentParentIds;

  String? labelFor(String id) => labels[id];
}

String _cap(String s) => s.isEmpty ? s : s[0].toUpperCase() + s.substring(1);

/// All ancestors of [id] (excluding self), via parent edges.
Set<String> _ancestors(FamilyGraph g, String id) {
  final out = <String>{};
  final queue = <String>[id];
  while (queue.isNotEmpty) {
    final cur = queue.removeAt(0);
    for (final p in g.parentsOf[cur] ?? const <String>[]) {
      if (out.add(p)) queue.add(p);
    }
  }
  return out;
}

/// All descendants of [id] (excluding self), via child edges.
Set<String> _descendants(FamilyGraph g, String id) {
  final out = <String>{};
  final queue = <String>[id];
  while (queue.isNotEmpty) {
    final cur = queue.removeAt(0);
    for (final c in g.childrenOf[cur] ?? const <String>[]) {
      if (out.add(c)) queue.add(c);
    }
  }
  return out;
}

Lineage computeLineage(FamilyGraph g, String selectedId, {bool full = false}) {
  final parents = {...?g.parentsOf[selectedId]};
  final spouses = {...?g.spousesOf[selectedId]};
  final children = {...?g.childrenOf[selectedId]};

  // Siblings = other children of the selected person's parents.
  final siblings = <String>{};
  for (final p in parents) {
    for (final c in g.childrenOf[p] ?? const <String>[]) {
      if (c != selectedId) siblings.add(c);
    }
  }

  final Set<String> descentChildIds;
  final Set<String> descentParentIds;
  final Set<String> related;

  if (full) {
    final ancestors = _ancestors(g, selectedId);
    final descendants = _descendants(g, selectedId);
    // Each ancestor's up-link + the selected person's own up-link.
    descentChildIds = {selectedId, ...ancestors, ...siblings};
    // The selected person + every descendant as a parent of their children.
    descentParentIds = {selectedId, ...descendants};
    related = {...parents, ...siblings, ...spouses, ...children,
      ...ancestors, ...descendants};
  } else {
    descentChildIds = {selectedId, ...siblings};
    descentParentIds = {selectedId};
    related = {...parents, ...siblings, ...spouses, ...children};
  }

  final members = {selectedId, ...related};
  final unionMembers = full ? members : {selectedId, ...spouses};

  final labels = <String, String>{};
  for (final id in related) {
    final k = computeKinship(g, selectedId, id);
    if (k.isTerm) labels[id] = _cap(k.label);
  }

  return Lineage(
    selectedId: selectedId,
    members: members,
    labels: labels,
    unionMembers: unionMembers,
    descentChildIds: descentChildIds,
    descentParentIds: descentParentIds,
  );
}

File: webapp/src/lib/lineage.ts
import type { Member, Relationship } from './types'
import { shortestPath } from './relationships'
import { describeKinship } from './kinship'

export interface Lineage {
  selectedId: string
  members: Set<string> // selected + all highlighted relatives
  labels: Map<string, string> // relativeId -> capitalized relationship
  unionMembers: Set<string> // spouse bar highlights if either end here
  descentChildIds: Set<string> // a child's up-link highlights if child here
  descentParentIds: Set<string> // a parent's down-links highlight if parent here
}

interface Graph {
  parentsOf: Map<string, string[]>
  childrenOf: Map<string, string[]>
  spousesOf: Map<string, string[]>
}

function buildGraph(members: Member[], rels: Relationship[]): Graph {
  const parentsOf = new Map<string, string[]>()
  const childrenOf = new Map<string, string[]>()
  const spousesOf = new Map<string, string[]>()
  for (const m of members) {
    parentsOf.set(m.id, [])
    childrenOf.set(m.id, [])
    spousesOf.set(m.id, [])
  }
  for (const r of rels) {
    if (r.type === 'parent') {
      childrenOf.get(r.from_member)?.push(r.to_member)
      parentsOf.get(r.to_member)?.push(r.from_member)
    } else {
      spousesOf.get(r.from_member)?.push(r.to_member)
      spousesOf.get(r.to_member)?.push(r.from_member)
    }
  }
  return { parentsOf, childrenOf, spousesOf }
}

function ancestors(g: Graph, id: string): Set<string> {
  const out = new Set<string>()
  const queue = [id]
  while (queue.length) {
    const cur = queue.shift()!
    for (const p of g.parentsOf.get(cur) ?? []) if (!out.has(p)) { out.add(p); queue.push(p) }
  }
  return out
}
function descendants(g: Graph, id: string): Set<string> {
  const out = new Set<string>()
  const queue = [id]
  while (queue.length) {
    const cur = queue.shift()!
    for (const c of g.childrenOf.get(cur) ?? []) if (!out.has(c)) { out.add(c); queue.push(c) }
  }
  return out
}

const cap = (s: string) => (s ? s[0].toUpperCase() + s.slice(1) : s)

/** Highlight sets for "View lineage" — mirrors the Android computeLineage. */
export function computeLineage(
  members: Member[],
  rels: Relationship[],
  selectedId: string,
  full = false,
): Lineage {
  const g = buildGraph(members, rels)
  const parents = new Set(g.parentsOf.get(selectedId) ?? [])
  const spouses = new Set(g.spousesOf.get(selectedId) ?? [])
  const children = new Set(g.childrenOf.get(selectedId) ?? [])

  const siblings = new Set<string>()
  for (const p of parents) {
    for (const c of g.childrenOf.get(p) ?? []) if (c !== selectedId) siblings.add(c)
  }

  let descentChildIds: Set<string>
  let descentParentIds: Set<string>
  let related: Set<string>

  if (full) {
    const anc = ancestors(g, selectedId)
    const desc = descendants(g, selectedId)
    descentChildIds = new Set([selectedId, ...anc, ...siblings])
    descentParentIds = new Set([selectedId, ...desc])
    related = new Set([...parents, ...siblings, ...spouses, ...children, ...anc, ...desc])
  } else {
    descentChildIds = new Set([selectedId, ...siblings])
    descentParentIds = new Set([selectedId])
    related = new Set([...parents, ...siblings, ...spouses, ...children])
  }

  const membersSet = new Set([selectedId, ...related])
  const unionMembers = full ? membersSet : new Set([selectedId, ...spouses])

  const labels = new Map<string, string>()
  for (const id of related) {
    const path = shortestPath(selectedId, id, rels)
    if (path && path.length) labels.set(id, cap(describeKinship(path)))
  }

  return { selectedId, members: membersSet, labels, unionMembers, descentChildIds, descentParentIds }
}
